### PR TITLE
impl(GCS+gRPC): resumable uploads wrapper

### PR DESCRIPTION
**Pull request by coryan**
Introduce `storage_experimental::AsyncWriter`. This class plays the `*Client` role for resumable uploads. It has a less error-prone interface than `storage_experimental::AsyncWriterConnection`, with all rules about which calls can be made concurrently enforced by the API and/or runtime errors.

Part of the work for #9134

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13008)

### CODE REVIEW

File: google/cloud/storage/async_writer.h [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+namespace cloud {
+namespace storage_experimental {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * Perform resumable uploads asynchronously.
+ *
+ * Resumable uploads allow applications to continue uploading data even after
+ * disconnects and even after application restarts. To resume an upload the
+ * library first queries the current state of the upload, the upload uses this
```
Comment (dbolduc) [inline google/cloud/storage/async_writer.h:38]: s/, the/. The/

File: google/cloud/storage/async_writer.h [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+ * information to send the remaining data. Applications only need to checkpoint
+ * a string, the `UploadId()`, to resume an upload even after the application
+ * itself restarts.
+ *
+ * Some data sources do not permit rewinding to an arbitrary point, for example:
+ * if the application is receiving streaming data from an external sourcein it
```
Comment (dbolduc) [inline google/cloud/storage/async_writer.h:44]: s/sourcein/source/

File: google/cloud/storage/async_writer.h [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+ * may be impossible to shutdown the application and recover the data streamed
+ * while the application was down.
+ *
+ * Nevertheless, even with streaming data sources, resumable uploads can be used
+ * to recover from most network problems between the application and Google
+ * Cloud Storage. The application can buffer any streaming data in memory, and
+ * periodically force a flush in the resumable upload. The service will return
+ * the number of bytes persisted and the application and then discard the
```
Comment (dbolduc) [inline google/cloud/storage/async_writer.h:52]: s/application and/application can/ ?

File: google/cloud/storage/async_writer.h [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+ * corresponding bytes from its buffer.
+ *
+ * Some caveats apply:
+ * - The application should be prepared for "partial flush" results, that is
```
Comment (dbolduc) [inline google/cloud/storage/async_writer.h:56]: s/, that is/. That is, the/

File: google/cloud/storage/async_writer.h [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+ *   service may indicate than fewer bytes were persisted vs. the number of
```
Comment (dbolduc) [inline google/cloud/storage/async_writer.h:57]: s/than/that/

File: google/cloud/storage/async_writer.h [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+   *
+   * This returns the last "known" state of the upload. The values are updated
+   * (1) when the `AsyncWriter` object is created, and (2) on calls to `Query()`
+   * and `Finalize()`.
+   *
+   * Applications may finalize an upload, and then try to resume the upload. In
+   * this case this function returns `storage::ObjectMetadata`.
+   *
+   * More commonly, applications resume an upload that is in progress, or
+   * periodically flush the
```
Comment (dbolduc) [inline google/cloud/storage/async_writer.h:96]: flush the buffer?

File: google/cloud/storage/async_writer.h [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+   *
+   * During an upload the service will periodically persist the uploaded data.
+   * Applications should not assume that all data "sent" is persisted. The
+   * service may buffer the data in ephemeral storage before persisting it.
+   *
+   * If the application needs to resume an upload that was interrupted, then it
+   * must check `persisted_state()` before sending data. The service will assume
```
Comment (dbolduc) [inline google/cloud/storage/async_writer.h:103]: s/persisted_state/PersistedState/

File: google/cloud/storage/async_writer.h [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+   * that new data starts at the correct offset.
+   *
+   * If an upload is resumed after it is finalized the library will return a
```
Comment (dbolduc) [inline google/cloud/storage/async_writer.h:106]: s/finalized/finalized,/

File: google/cloud/storage/async_writer.h [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+  absl::variant<std::int64_t, storage::ObjectMetadata> PersistedState() const;
+
+  /**
+   * Returns the offset for the next `Write()` or `Finalize()` call.
+   *
+   * This is automatically updated once the `Write()` call is satisfied.
+   * Applications may use this information to seek the right data from the
+   * source.
+   */
+  absl::optional<std::int64_t> current_offset() const;
```
Comment (dbolduc) [inline google/cloud/storage/async_writer.h:122]: `CurrentOffset()` ?

File: google/cloud/storage/async_writer_test.cc [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        return make_ready_future(make_status_or(storage::ObjectMetadata()));
+      });
+
+  auto token = storage_internal::MakeAsyncToken(mock.get());
+  AsyncWriter writer(std::move(mock));
+  EXPECT_EQ(writer.UploadId(), "test-upload-id");
+  EXPECT_EQ(absl::get<std::int64_t>(writer.PersistedState()), 16384);
+  token = writer.Write(std::move(token), WritePayload{std::string("aaa")})
+              .get()
+              .value();
```
Comment (dbolduc) [inline google/cloud/storage/async_writer_test.cc:64]: shouldn't we `ASSERT_STATUS_OK()` before using the value?

File: google/cloud/storage/async_writer_test.cc [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+  });
+
+  auto token = storage_internal::MakeAsyncToken(mock.get());
+  AsyncWriter writer(std::move(mock));
+  auto const actual = writer.Finalize(std::move(token), WritePayload{}).get();
+  EXPECT_THAT(actual, StatusIs(PermanentError().code()));
+}
+
+TEST(AsyncWriterTest, ErrorOnClosedStream) {
+  AsyncWriter writer;
```
Comment (dbolduc) [inline google/cloud/storage/async_writer_test.cc:122]: I am a little confused by the "Closed Stream" thing. Why do we even have a default constructor?

What should happen with `UploadId()` / `PersistedState()`? They do not handle errors `if (!impl)`.

File: google/cloud/storage/async_writer.h [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+namespace google {
+namespace cloud {
+namespace storage_experimental {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * Perform resumable uploads asynchronously.
+ *
+ * Resumable uploads allow applications to continue uploading data even after
+ * disconnects and even after application restarts. To resume an upload the
```
Comment (devbww) [inline google/cloud/storage/async_writer.h:37]: s/ even after//

File: google/cloud/storage/async_writer.h [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+ * library first queries the current state of the upload, the upload uses this
+ * information to send the remaining data. Applications only need to checkpoint
+ * a string, the `UploadId()`, to resume an upload even after the application
+ * itself restarts.
+ *
+ * Some data sources do not permit rewinding to an arbitrary point, for example:
```
Comment (devbww) [inline google/cloud/storage/async_writer.h:43]: s/, for example:/. For example,/

File: google/cloud/storage/async_writer.h [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+ * the number of bytes persisted and the application and then discard the
+ * corresponding bytes from its buffer.
+ *
+ * Some caveats apply:
+ * - The application should be prepared for "partial flush" results, that is
+ *   service may indicate than fewer bytes were persisted vs. the number of
+ *   bytes sent.
+ * - The application can continue to upload data after a flush. It is possible
+ *   that a flush request returns more persisted bytes than were available when
+ *   the flush request is sent.
```
Comment (devbww) [inline google/cloud/storage/async_writer.h:61]: s/is/was/

File: google/cloud/storage/async_writer.h [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+   * application should send the remaining data to upload, starting from this
+   * point.
+   */
+  absl::variant<std::int64_t, storage::ObjectMetadata> PersistedState() const;
+
+  /**
+   * Returns the offset for the next `Write()` or `Finalize()` call.
+   *
+   * This is automatically updated once the `Write()` call is satisfied.
+   * Applications may use this information to seek the right data from the
```
Comment (devbww) [inline google/cloud/storage/async_writer.h:119]: I'm not sure what "to seek the right data" means.

File: google/cloud/storage/async_writer.cc [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+      internal::InvalidArgumentError("invalid token", std::move(eib))));
+}
+
+template <typename T>
+future<StatusOr<T>> StreamError(internal::ErrorInfoBuilder eib) {
+  return make_ready_future(
+      StatusOr<T>(internal::CancelledError("closed stream", std::move(eib))));
+}
+
+using FlushResponse = std::pair<AsyncToken, future<StatusOr<std::int64_t>>>;
```
Comment (devbww) [inline google/cloud/storage/async_writer.cc:36]: Unused?

Comment (coryan) [reply google/cloud/storage/async_writer.cc]: Removed

File: google/cloud/storage/async_writer.h (hunk 15 above)
Comment (coryan) [reply google/cloud/storage/async_writer.h]: Removed for now

File: google/cloud/storage/async_writer.h (hunk 14 above)
Comment (coryan) [reply google/cloud/storage/async_writer.h]: Removed for now

File: google/cloud/storage/async_writer.h (hunk 13 above)
Comment (coryan) [reply google/cloud/storage/async_writer.h:42]: Removed for now

File: google/cloud/storage/async_writer.h (hunk 12 above)
Comment (coryan) [reply google/cloud/storage/async_writer.h]: Fixed.

File: google/cloud/storage/async_writer_test.cc (hunk 11 above)
Comment (coryan) [reply google/cloud/storage/async_writer_test.cc:122]: > I am a little confused by the "Closed Stream" thing. Why do we even have a default constructor?

You need it to support `std::tie()`, to put these into `std::pair` and it is the result of `std::move()`-from.

> 
> What should happen with `UploadId()` / `PersistedState()`? They do not handle errors `if (!impl)`.

Documented that as "undefined behavior".  The fact that I picked a specific behavior does not contradict the documentation.

File: google/cloud/storage/async_writer_test.cc (hunk 10 above)
Comment (coryan) [reply google/cloud/storage/async_writer_test.cc:64]: I think we are fine: `.value()` throws and stops the test too.

File: google/cloud/storage/async_writer.h (hunk 9 above)
Comment (coryan) [reply google/cloud/storage/async_writer.h]: Removed.

File: google/cloud/storage/async_writer.h (hunk 7 above)
Comment (coryan) [reply google/cloud/storage/async_writer.h]: Fixed.

File: google/cloud/storage/async_writer.h (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/async_writer.h]: Done.

File: google/cloud/storage/async_writer.h (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/async_writer.h]: Done.

File: google/cloud/storage/async_writer.h (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/async_writer.h]: Done.

File: google/cloud/storage/async_writer.h (hunk 5 above)
Comment (coryan) [reply google/cloud/storage/async_writer.h]: Removed for now.

File: google/cloud/storage/async_writer.h (hunk 4 above)
Comment (coryan) [reply google/cloud/storage/async_writer.h]: I am removing this section.  It needs member functions that will appear in a future PR.

File: google/cloud/storage/async_writer.h [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+ * `UploadId()`, to resume an upload even after the application itself restarts.
+ *
+ * Some data sources do not permit rewinding to an arbitrary point. For example:
+ * if the application is receiving streaming data from an external source it may
+ * be impossible to shutdown the application and recover the data streamed while
+ * the application was down.
+ *
+ * This API does not support resuming uploading data from streaming data
+ * sources. If the upload is interrupted you must be able to start sending data
+ * from an arbitrary point
```
Comment (devbww) [inline google/cloud/storage/async_writer.h:49]: s/$/./

**Review by devbww — APPROVED** [review]

File: google/cloud/storage/async_writer.h (hunk 17 above)
Comment (coryan) [reply google/cloud/storage/async_writer.h]: Fixed, thanks.

**Pull request merged**